### PR TITLE
Allow changing mode in navigationOptions for dynamic modal

### PR DESCRIPTION
**Pull request by JulianKingman**
I need dynamic modals in my project and couldn't wait for https://github.com/react-community/react-navigation/pull/1378. It looked simple enough to implement, so I went ahead with it :)

I've got navigation animations working, but I'm having a heck of a time configuring the "back" animation.

Any idea where I can configure the navigator in such a way that I can influence the animation, and have per-screen navigationOptions available? As far as I can tell, `isModal` is currently defined in `CardStackTransitioner` here: https://github.com/react-community/react-navigation/blob/master/src/views/CardStack/CardStackTransitioner.js#L77, which seems to apply isModal to the navigator universally.

### CODE REVIEW

**Comment by JulianKingman** [comment]
Turns out I was wrong, setting isModal in CardStackTransitioner doesn't do the trick either, so I have no idea where the back animation comes from if not there, I can't find any reference to it.

**Comment by davojan** [comment]
@JulianKingman Back animation seems to be a separate issue. Even without your changes it's behaviour is odd in "Modal Stack Example" - "Go to profile screen" is animated vertically, but back-animation is horizontal.

In your modified example back navigation starts to work as expected if you click "Go to a photos screen" one more deeper and then click "Go back".

**Comment by davojan** [comment]
Nope, I was wrong. In "Modal Stack Example" there is a nested navigator that causing unintuitive behaviour.

But the good news is I made it to work correctly during back navigation! Check this out:
https://github.com/procraft/react-navigation/commit/fa2a383fde51d76e0b05edee5b56d3c590b36a55#diff-95db94c88696779a7ac6a0fe1af87ebeR390

**Comment by JulianKingman** [comment]
Brilliant, thanks @davojan! I added the fix to my PR.

**Comment by JulianKingman** [comment]
- **Test thoroughly and describe how you tested it**
I added examples to  the NavigationPlayground and tested a modal type screen in a card stack, and a card type screen in a modal
- **Add an example to the playground app that utilizes this new option**
Yes, done (see SimpleStack and ModalStack for examples)
- **Take screenshots/videos of the behavior on both platforms**
Here: 
modalnavigationoption
- **Write documentation**
Done

**Comment by JulianKingman** [comment]
@ericvicenti FYI, ready for review

**Comment by davojan** [comment]
@JulianKingman I guess you should fix linter errors reported by the circleci before they'll consider reviewing :)

**Comment by JulianKingman** [comment]
Fixed.

**Comment by claykohut** [comment]
Getting this error when trying to use this fork:

screen shot 2017-08-31 at 1 00 27 pm

**Comment by maxstoller** [comment]
Working well for me, would be nice to see it merged. Thanks for the contribution, @JulianKingman!

**Comment by kelset** [comment]
Looks like there are some conflicts, can you fix them? I'll try to check the PR asap

File: examples/NavigationPlayground/package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
     "flow-bin": "^0.49.1",
     "react": "16.0.0-alpha.12",
     "react-native": "^0.45.1",
-    "react-navigation": "*"
+    "react-navigation": "designmanio/react-navigation"
```
Comment (charpeni) [inline examples/NavigationPlayground/package.json]: It looks like an artifact?

**Comment by dccarmo** [comment]
Any updates on this? It seems the conflicts are minimal (documentation and dependencies), so I guess it shouldn't be too hard to update and approve it?

**Comment by claykohut** [comment]
id also love to get this merged

**Comment by JulianKingman** [comment]
So I fixed the conflicts, but the examples appear to be broken, and I didn't want to go in and change everything to get them working.

**Comment by kelset** [comment]
Ok, thanks. I'll take a look at it this weekend, but I'm not sure when I'll merge it. We're waiting on a big PR to fix a lot of the codebase and I'd prefer to land that one before this; we'll see.

**Comment by JulianKingman** [comment]
This PR touches a few parts, so smart move, thanks!

**Comment by antgly** [comment]
Are you still attempting to tackle this @kelset?

**Comment by kelset** [comment]
@antgly as I mentioned in my previous comment, we first had to merge some huge PRs (that become `beta.16`) before thinking of merging this. We are now tackling a couple issues related to the new version.

That said I won't review this personally but it's in a shortlist of PRs we are trying to review & merge.

**Comment by rborn** [comment]
Could this be merged ? how can I help ?
Thnx

**Comment by kelset** [comment]
@rborn 
> Could this be merged ?
No, at the moment there are some conflicts that must be resolved.

> how can I help ?
You can add your review, which will surely help us.

File: src/views/CardStack/CardStackTransitioner.js [hunk 2]
```diff
@@ -2,11 +2,14 @@
 
 import React, { Component } from 'react';
 import { NativeModules } from 'react-native';
+import _ from 'lodash';
```
Comment (rborn) [inline src/views/CardStack/CardStackTransitioner.js]: Do we use lodash or it's a ghost import :) ?

File: src/views/CardStack/CardStackTransitioner.js [hunk 3]
```diff
@@ -2,11 +2,14 @@
...
 
 import CardStack from './CardStack';
 import CardStackStyleInterpolator from './CardStackStyleInterpolator';
 import Transitioner from '../Transitioner';
+import addNavigationHelpers from '../../addNavigationHelpers';
 import TransitionConfigs from './TransitionConfigs';
+import getScreenForRouteName from '../../routers/getScreenForRouteName';
```
Comment (rborn) [inline src/views/CardStack/CardStackTransitioner.js]: Same for getScreenForRouteName ?

File: src/views/CardStack/CardStackTransitioner.js [hunk 4]
```diff
@@ -74,7 +77,9 @@ class CardStackTransitioner extends Component<DefaultProps, Props, void> {
     // props for the old screen
     prevTransitionProps: NavigationTransitionProps
   ) => {
+    // const isModal = this._getMode() === 'modal';
     const isModal = this.props.mode === 'modal';
+
```
Comment (rborn) [inline src/views/CardStack/CardStackTransitioner.js]: do we need the extra line ?

File: src/views/CardStack/CardStackTransitioner.js [hunk 5]
```diff
@@ -74,7 +77,9 @@ class CardStackTransitioner extends Component<DefaultProps, Props, void> {
     // props for the old screen
     prevTransitionProps: NavigationTransitionProps
   ) => {
+    // const isModal = this._getMode() === 'modal';
```
Comment (rborn) [inline src/views/CardStack/CardStackTransitioner.js]: Is this comment any relevant here?

**Comment by rborn** [comment]
@kelset am I blind or the reported conflicting  files don't really have conflicts?

**Comment by kelset** [comment]
@rborn first off thanks for the review. Second, it's weird, I can see the conflicting sections.

**Comment by rborn** [comment]
@kelset ignore me, I'm dumb :|

**Comment by Andreyco** [comment]
Does it mean, we will no longer need separate "modal" navigators?

**Comment by rborn** [comment]
@Andreyco I think yes, that's the idea. This PR would simplify a lot the navigation :). 
 @JulianKingman correct?

**Comment by JulianKingman** [comment]
Correct. This PR allows a `mode` option in navigationOptions, so it can be set at the router or route level.

**Comment by rborn** [comment]
@JulianKingman lovely, could you have a look at the review and fix the conflicts, and the maybe we can bribe @kelset to merge it? 😋

File: src/views/CardStack/CardStackTransitioner.js [hunk 6]
```diff
@@ -62,7 +65,9 @@ class CardStackTransitioner extends React.Component<Props> {
     // props for the old screen
     prevTransitionProps: ?NavigationTransitionProps
   ) => {
+    // const isModal = this._getMode() === 'modal';
```
Comment (spencercarli) [inline src/views/CardStack/CardStackTransitioner.js]: Hey @JulianKingman I'm confused by this line. Should this be used or not? If not we should remove this.

File: src/views/CardStack/CardStackTransitioner.js [hunk 7]
```diff
@@ -84,11 +89,27 @@ class CardStackTransitioner extends React.Component<Props> {
...
   };
 
+  _getMode() {
+    const { navigation, mode, router } = this.props;
+    let screenOptions;
+    if (
+      this.props.navigation.state.routes &&
+      this.props.navigation.state.routes[1]
+    ) {
+      router.getScreenOptions(
```
Comment (spencercarli) [inline src/views/CardStack/CardStackTransitioner.js]: Shouldn't we be storing this on `screenOptions`?

**Comment by welkinwong** [comment]
i like this

**Comment by Andreyco** [comment]
One question here... Is it possible to override _mode_ for screen when it is pushed?
Imagine situation where you have, e.g, Profile screen. In one case, you want to push it as card (from right to left), in another case you need modal.

Is something like this possible?
```js
props.navigation.navigate('ProfileScreen', { mode: 'modal' })
```

**Comment by stereodenis** [comment]
@Andreyco maybe you can use it in navigationOptions in Profile component

**Comment by stereodenis** [comment]
@JulianKingman could you update your branch to latest master?

**Comment by Andreyco** [comment]
> maybe you can use it in navigationOptions in Profile component

If I am not mistaken, unless nav options are not passed at "push", they are could be consider as "static" a have to be defined prior to "push".

In my opinion, how is screen presented could/should be defined dynamically. It's other navigation options (title, buttons, ...) could be defined upfront (as is now)

**Comment by helielson** [comment]
@JulianKingman good job.
Are you having any issues to update this branch?

File: src/views/CardStack/CardStackTransitioner.js (hunk 6 above)
Comment (JulianKingman) [reply src/views/CardStack/CardStackTransitioner.js]: I wasn't sure if it was working, but preferred using the getMode() function. It works now.

File: src/views/CardStack/CardStackTransitioner.js (hunk 7 above)
Comment (JulianKingman) [reply src/views/CardStack/CardStackTransitioner.js]: Yes, I fixed it to make more sense.

File: .expo/settings.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+  "hostType": "tunnel",
+  "lanType": "ip",
+  "dev": true,
+  "strict": false,
+  "minify": false,
+  "urlType": "exp",
+  "urlRandomness": null
+}
```
Comment (helielson) [inline .expo/settings.json]: @JulianKingman is this file needed?

File: examples/NavigationPlayground/js/ModalStack.js [hunk 9]
```diff
@@ -14,32 +14,44 @@ const MyNavScreen = ({ navigation, banner }) => (
         top: navigation.state.routeName === 'HeaderTest' ? 'always' : 'never',
       }}
     >
-      <SampleText>{banner}</SampleText>
+    <SampleText>{banner}</SampleText>
```
Comment (helielson) [inline examples/NavigationPlayground/js/ModalStack.js]: @JulianKingman Fix the indentation of SafeAreaView's children components

File: examples/NavigationPlayground/js/SimpleStack.js [hunk 10]
```diff
@@ -40,13 +40,28 @@ class MyNavScreen extends React.Component<MyNavScreenProps> {
...
         <Button onPress={() => navigation.goBack(null)} title="Go back" />
         <StatusBar barStyle="default" />
       </SafeAreaView>
     );
   }
 }
 
+const Modal = ({navigation}) => (
+  <MyNavScreen
+    banner={`Modal in Screen Navigator`}
```
Comment (helielson) [inline examples/NavigationPlayground/js/SimpleStack.js]: Strings should use singlequote.
`banner="Modal in Screen Navigator"`

File: examples/NavigationPlayground/package.json [hunk 11]
```diff
@@ -14,7 +14,7 @@
     "expo": "^25.0.0",
     "react": "16.2.0",
     "react-native": "^0.52.0",
-    "react-navigation": "link:../.."
+    "react-navigation": "file:../.."
```
Comment (helielson) [inline examples/NavigationPlayground/package.json]: @JulianKingman is it correct?

Comment (JulianKingman) [reply examples/NavigationPlayground/package.json]: @helielson npm doesn't support `link:`, so is yarn-only OK? If so, I'll change it back to `link:`
more here: https://github.com/codaxy/cxjs/issues/279#issuecomment-335096573

File: examples/NavigationPlayground/js/ModalStack.js (hunk 9 above)
Comment (JulianKingman) [reply examples/NavigationPlayground/js/ModalStack.js]: fixed

File: .expo/settings.json (hunk 8 above)
Comment (JulianKingman) [reply .expo/settings.json]: @helielson that's an expo file, which should be ignored. I updated the gitignore (expo directory was the outdated `.exponent`, instead of `.expo`), and removed it.

File: examples/NavigationPlayground/package.json (hunk 11 above)
Comment (helielson) [reply examples/NavigationPlayground/package.json]: indeed! 👍

**Comment by helielson** [comment]
@spencercarli @stereodenis @Andreyco @kelset the branch is now updated.
Can you review it again?

File: examples/NavigationPlayground/package.json (hunk 11 above)
Comment (JulianKingman) [reply examples/NavigationPlayground/package.json]: @helielson I changed it back to link, as the docs (https://reactnavigation.org/docs/guides/contributors#Run-the-Example-App) suggest using yarn.

**Comment by kelset** [comment]
I think @ericvicenti should be the one reviewing this because we are undergoing some massive changes so I'm not sure this will be needed

**Comment by raarts** [comment]
> we are undergoing some massive changes

is that documented somewhere, this makes me worried.

**Comment by kelset** [comment]
You can check the commits on master. We are not yet ready release a new version / post the release notes because we are still changing stuff, but yeah PR section & commits in master are the best way to be up-to-date if you don't want to wait for next release.

File: src/views/CardStack/CardStackTransitioner.js [hunk 12]
```diff
@@ -53,11 +54,26 @@ class CardStackTransitioner extends React.Component {
     return transitionSpec;
   };
 
+  _getMode() {
+    const { navigation, mode, router } = this.props;
+    // If there's a route, get its screen options
+    if (navigation.state.routes && navigation.state.routes[1]) {
+      screenOptions = router.getScreenOptions(
+        addNavigationHelpers({
+          state: navigation.state.routes[1],
```
Comment (ericvicenti) [inline src/views/CardStack/CardStackTransitioner.js]: It looks like we will only check the options of routes[1], which is the second screen in the stack. What if the third route is configured to be a modal? Can you try this out in the navigation playground and possibly post a video of what it looks like?

**Comment by ericvicenti** [comment]
And yes, as @kelset  mentioned, we do have some big changes happening in this area. It should make features like this easier to add, moving forward!

See the suggested changes here: https://github.com/react-navigation/react-navigation/pull/3392
and the RFC here: https://github.com/react-navigation/rfcs/blob/master/text/0002-navigator-view-api.md

@raarts , we are ramping up support, it shouldn't concern you! Feel free to follow along with progress on the issues and PRs in this repo, and by keeping an eye on the RFCs repo: https://github.com/react-navigation/rfcs

**Comment by RobPando** [comment]
My bad for the intrusion, but is this still in consideration to merge in or similar future solution? I need this bad.

**Comment by brentvatne** [comment]
I agree with @ericvicenti's assessment here, I'm not sure that this is the way we want to go for supporting modals. It may be, but I'd like to discuss further and have a proper RFC for this as it's a really key piece of the ergonomics of a navigation library (that we have up until now not provided a well-defined and supported path for, aside from creating a modal stack navigator). Please see this RFC placeholder issue: https://github.com/react-navigation/rfcs/issues/15 -- if you want to take a stab at designing a solution to this, let's discuss there and we can take it into a formal RFC. Thank you @JulianKingman for this PR and creating good discussion around this, and sorry to everyone who has to continue to use a fork to support this feature!
